### ISSUE TITLE
Add Script property to InteractiveScriptGlobals that returns a ScriptEnvironment object

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn#5979]
Move Args from InteractiveScriptGlobals to this object.

```c#
public class InteractiveScriptGlobals 
{
   public ScriptEnvironment Script { get; } = new ScriptEnvironment(...);
}

public class ScriptEnvironment 
{
    IList<string> Args { get; }
    string GetPath([System.Runtime.CompilerServices.CallerFilePath]string path = null) => path;
    string GetDirectory([System.Runtime.CompilerServices.CallerFilePath]string path = null) => Path.GetDirectory(path);
}
```

Post U1 we can add more features to this environment:

```c#
class ScriptEnvironment 
{
    string Language { return "CSharp" : "VB" }
    Task<object> EvalAsync(string code, object globals);
}
```

**Comment by tmat** [comment]
@ManishJayaswal @dotnet/interactive @mattwar

**Comment by tmat** [comment]
@MadsTorgersen Pointed out that Args are likely to be accessed frequently, so we might want to keep it directly on the Globals object. It would also avoid the breaking change.

**Comment by khellang** [comment]
Here are some things we (almost) added to scriptcs a couple of years back; https://github.com/scriptcs/scriptcs/pull/404

**Comment by tmat** [comment]
@khellang Thanks for the pointer. I'd be very reluctant to add any Assembly information. That just exposes implementation details of the host and asks for trouble. What was the scenario that it was enabling?

**Comment by khellang** [comment]
It's from https://github.com/scriptcs/scriptcs/issues/244. It's basically to be able to do assembly scanning etc.

**Comment by tmat** [comment]
I'd need to see more detailed reasoning than "I need it for Nancy".

**Comment by khellang** [comment]
> Currently I'm doing Assembly.GetCallingAssembly(). Whilst this works, it feels fragile and it's not discoverable.

A lot of existing APIs expect you to pass it an assembly, for whatever reason. It would be nice to have a convenient way of accessing it. Is an `Assembly` method the recommended way of doing it?

**Comment by tmat** [comment]
It depends on what such APIs expect to find in the assembly. Could you give examples of the APIs? We can then recommend how to use them.

It could be that these APIs are not well designed and can't be reliably used in scripts (are these APIs usable in IronPython, or other dynamic languages?). You can perhaps come up with a workaround for each APIs, but exposing assembly as a generally applicable solution would be wrong. It would give the impression that it works in general, but it won't.

**Comment by khellang** [comment]
Well, the most common example I can think of is scanning for types, e.g. to
1. Discover web framework controllers/modules. (Granted, these usually work in an AppDomain scope, but still)
2. Register things in an IoC container. http://docs.autofac.org/en/latest/register/scanning.html#scanning-for-types
3. Gather entity configurations (with EF). https://msdn.microsoft.com/en-us/library/system.data.entity.modelconfiguration.configuration.configurationregistrar.addfromassembly(v=vs.113).aspx

There are loads and loads of these examples. I guess you could always do something like `typeof(T).Assembly`, but still, it would be nice to have access to the "script" assembly somewhere :smile:

**Comment by glennblock** [comment]
As @khellang said this is pretty common. One use case I have used this for is controllers for Web API where the controllers themselves are authored within a script. The assembly is passed to an Assembly Resolver which reflects through the in-memory compiled assembly to get the controllers. Another common example would be a unit-test runner that executes tests within a script.

**Comment by tmat** [comment]
It seems to me that most of these scenarios would be addressed by providing ```IEnumerable<Type> Script.GetDeclaredTypes()```. Correct?

**Comment by khellang** [comment]
> It seems to me that most of these scenarios would be addressed by providing IEnumerable<Type> Script.GetDeclaredTypes(). Correct?

Yes. That would cover a lot of scenarios :smile:

**Comment by glennblock** [comment]
Yes that should work, if one needs the assembly they can grab it from the
type.

**Comment by glennblock** [comment]
Thanks @tmat

**Comment by tmat** [comment]
@glennblock The reason why I don't want to expose assembly directly is because scripts may be compiled into multiple assemblies (e.g. in REPL) and we don't want different behavior between REPL and scripts.

GetDeclaredTypes() will also handle type hiding in REPL.

```C#
> class C { public int X; }
> class C { public int Y; }
> new C()
C { X = 0, Y = 0 }
```

GetDeclaredTypes() will return the last defined C, not the previous ones.

**Comment by glennblock** [comment]
Nice, thanks for the clarification. Most frameworks that I know of that
accept assemblies also accept type collections so this should work.

**Comment by ManishJayaswal** [comment]
@tmat moving this to 1.2 milestone
